drivers/clock_control/clock_control_ll_nrfxlib.c

```diff
@@ -0,0 +1,150 @@
+/*
```

Comment:
I am really against having non-radio related drivers coupled to BLE stack. There is already a clock driver in the Zephyr, there is planned work which will change API of the clock management etc. If we merge this, the work needed to improve this area will double.
Moreover I see that ble_controller_* API is not compatible with the planned features (like callback when clock source changes). So after merge any improvement in the area will be basically blocked due to different priorities of the teams and incompatible API (we already observed that in nRF5 SDK).

Reply:
This has already been up for discussion and the concerns about duplicate work were raised, I believe @thomasstenersen has written down details of how this decision was made.

Reply:
I really think that it is not the right path. I still remember all problems created by SoftDevice design choices which were 100% OK when made, but later, when SW complexity increased, become really painful. And any attempt to change that was refused due to "priorities", "schedule" etc. Please note, that NCS is intended to improve cooperation between teams and help to solve this issue. If we will make the same mistakes, this will not succeed.

I do not see any benefit from closed-source clock management integrated with BLE stack (as far as I know there are no secrets we would like to protect in this area) but I see whole bunch of problems.

Also the "we need OS-agnostic drivers so we cannot use Zephyr ones" is not justifying crating a new set of the drivers:
We have the nRFX which is already fulfilling this requirement.

Reply:
IIRC, the requirement they have that is the killer is that as opposed to the SDK, they can not distribute source code, but must be able to distribute closed source binary versions of the drivers.

Where this requirement comes from I don't know, if it could be removed, it would be possible to use nRFX drivers I believe.

Please correct me if I'm wrong @thomasstenersen .

Reply:
As far as I know, the binary is needed for certification (@thomasstenersen please confirm).
In such case the nRFX could be compiled and distributed as library when such form is required.

The key benefit from using that layer is that OS-related stuff (like API changes, DTS integration) is not affecting BLE as it does not need to export Zephyr API for the shared components as well as BLE changes are not affecting OS as our Zephyr drivers are built on top of the nRFX.

Reply:
I am currently away this week and cannot look into this in detail. Can you please organize a quick call among you 3 this week so as to try to find a solution?

Reply:
>  is needed for certification

I don't think that is correct. E.g. the Zephyr BLE controller is certified, even though it is not distributed as a binary.

Reply:
@pizi-nordic You raise a very valid point. The reason for adding these drivers now is not because that is the best solution or the solution we want, but a solution that we can get in right now.
We are moving away from the SoftDevice solution towards a new solution, but "Priorities" and "schedule" is still a problem for us right now, the changes you suggest are not trivial and requires time to implement and verify.
The SoftDevice team acknowledges that this is something that should be improved in the future, but the actual solution has not been agreed to yet. But that it needs to be improved is very much agreed upon.
Your suggestion to have ble_controller depend on nRFX has been suggested as one solution, if you have a very clear picture of how this could be done it would be very helpful if you could make a concrete suggestion to the SD team.

The question is if we can accept these drivers now as a temporary solution. As you say this requires double work to maintain. Is this acceptable for the short term, specifically for the next release?

Reply:
I am very afraid, that this temporal solution will be there forever due to "priorities", "schedule" and "quality" (why we should change something that works ,is tested, etc.). This fear comes from the struggling we already had with the SoftDevice. And here I see that exactly the same design, which brought us a lot of pain, is introduced again.

There is ongoing work in the NCS in clock control area. The API you proposed will not support usecases they require. I suppose that any change on the area will take a lot of time due to reasons mentioned above.

I really would like to change the design in the way that OS-related changes will not affect BLE and vice-versa. IMHO the biggest problem is direct coupling of non-BLE related stuff (like clock control) with BLE functionality. I cannot use the proposed clock control interface before or without BLE initialization. I am afraid that I will be not able to use it after BLE shutdown. I have no idea if my clocking settings (for example started HFCLK) will survive BLE initialization...

To move from (non constructive) complains to something we can discuss, here are the key things which IMHO should be changed looking from OS perspective:
- Clock control and entropy driver initialization/operation should be independent from BLE stack initialization/operation. 
- BLE stack should be regular consumer of the clock control. So if HFCLK was requested by the APP, it should not be stopped. This leads to the need of reference-count based management of the clock.
- Any changes of clock configuration (for example start/stop of given clock source) have to be notified to the OS/APP (via callback).

Reply:
For the last point, I would even add a note, that in (nearly) ideal design, BLE controller should request the clock source to be changed and leaving it to the application. I can imagine a design where I wish to enable BLE only once every day and upload the data. Then turn the BLE off, but I really cannot allow the BLE to disable my RTC as I would lose the timing then.

Reply:
> The question is if we can accept these drivers now as a temporary solution.

Will it be transparent for everybody using it? I.e. if I will switch from Zephyr LL to proposed solution how many changes we will need to do in the application? How much we will have to divert? So far I expected that when we select the new LL we will not be affected in any areas and application will work as is but with optional linked layer. From what is proposed it seems that there may be an impact.

Reply:
BTW - how will this work with zephyr using clocks for system tick and various drivers using them for running peripherals?

Reply:
Sorry for replying late, I've been out sick since Thursday. Thank you @joerchan
and @sebastianboe for providing some context.

@pizi-nordic, your concerns are certainly valid and I understand that the
situation is not ideal. However, this is not the time nor place for this
discussion. As @joerchan said, we would very much appreciate suggestions to
better the design, but let us do that internally and with concrete suggestions.
As we are still at a prealpha stage, the API is still considered volatile and
pushing as much of the low level drivers as possible out of the controller is
definitely something we want to achieve. The question is how to achieve it.
That being said, the proposed architecture has been reviewed across different
teams and it was decided that the current solution best meets the critera and
contraints we have. Also keep in mind that one of the primary design goals of
the controller is for it to be fully usable _without_ Zephyr/an RTOS.

We cannot do anything about the architecture for the time being, but will fix
any direct issues/problems with the PR. We acknowlegde that future changes to
Zephyr's clock management etc. will require updating two drivers and this is
a situation we don't want. Given that the drivers we provide in this PR only
implement the given APIs in Zephyr, changes to these drivers should not impact
user applications. Even if we completely change the interface between Zephyr and
the BLE controller. In short, changes will be needed either way -- let's find a
good way of handling this outside of this PR.

@pdunaj the drivers are implementing generic APIs in Zephyr, the application
should not be dependent on behavior in the underlying driver. There should be no
change to be made when switching LL's. However, the BT_LL_NRFXLIB is clearly
in a v0.2.0-1.prealpha and the user should expect that there will be breaking
changes coming in the controller library.

In conclusion, @pizi-nordic, @nvlsianpu, @pdunaj, et. al., are you saying that
the solution we've provided is not good enough for D4 and will not be accepted?
And as a consequence, the v0.2.0 of the Nordic proprietary BLE controller will
not be supported in NCS?

Reply:
Hi @thomasstenersen , thanks for the reply.

>  let's find a good way of handling this outside of this PR

I agree. I think this would be better to discuss internally.

> are you saying that the solution we've provided is not good enough for D4 and will not be accepted?

I consider myself very pragmatic. If we want this to go in then as long as it won't break anything else that we consider valuable it can be in. The question is: what will be value of doing that, especially it being part of the official release? Will any project be able to actually use the new LL in this shape? Or is this done mainly to see what is wrong and change/adapt code to Zephyr? If so why having it as part of the release?

If I am correct we will be one of the firsts to use this LL. The biggest question for me is: will this work at all with current Zephyr? Have you run this with any samples? If so which are working and which aren't? Do you have any report from integration and testing of new LL?

I propose to do following:
- grab the information from people around about required changes to architecture,
- run samples to see what does not work (if not done already),
- fix all the bugs clearly visible (leaving big arch reworks),
- make it clear that this does not fit zephyr well, plus point out what is not right (should be clearly seen after first and second points are done),
- have it integrated for others in NCS (us among them) to further tell what does not work so we can improve.

If this should block the D4?
Personally I don't think so, unless we know anybody will really use it. I.e. if you know it works on all samples and know people can start project development on it than it maybe worth waiting for. This is however my view on things.

Reply:
>The question is: what will be value of doing that
Yes. We have customers waiting :) Also this will make the LL actually workable. This PR along with the updated controller will fix a lot of outstanding issues.

> Will any project be able to actually use the new LL in this shape? 
Try the examples, they should pretty much all work out of the box. 

> Or is this done mainly to see what is wrong and change/adapt code to Zephyr? If so why having it as part of the release?
There are numerous reasons for this, and one of them is also to discover the pains of integration. Others are to get continuous builds of the LL out so we can minimize the friction for each update.

>  we will be one of the firsts to use this LL
I'm not sure who "we" are in this context, but this LL has been running smoothly for many, many years. What is new is how we try to integrate it with another host in an RTOS environment.

> make it clear that this does not fit zephyr well, plus point out what is not right (should be clearly seen after first and second points are done),
I'm not sure what kind of documentation/information you would like here and even less where to put it. I believe it is fairly clearly communicated that this is still in early development and it is not the default LL choice.

> have it integrated for others in NCS (us among them) to further tell what does not work so we can improve.
What does this mean?

In short, _yes_, we've tested with many examples and not found any LL related issues. I prose you try that as well :smile: 
My revised list of actions:

- [ ] Gather architectural feedback/required changes and distribute internally
- [X] Run samples
- [ ] Fix non-architectural feedback

Reply:
@thomasstenersen 

> In short, yes, we've tested with many examples and not found any LL related issues. I prose you try that as well smile

I am not sure if I am the best person to make sure this code works. Anyway I switched to this PR, took the first example from the row (`samples/bluetooth/peripheral_hids`) and got build failure after enabling `BT_LL_NRFXLIB`:
```
ubsys/bluetooth/controller/hci_driver.c:276:35: error: 'US32_MAX' undeclared (first use in this function); did you mean 'UINT_MAX'?
  if (((NRF_FICR->DEVICEADDR[0] != US32_MAX) ||
                                   ^~~~~~~~
                                   UINT_MAX
/home/pdunaj/work/ncs/nrf/subsys/bluetooth/controller/hci_driver.c:276:35
```

Fixed that, let it run and got following during runtime:
```
#23 0x00020100 in _Fault (esf=0x20009a38 <tx_thread_stack+496>, exc_return=<optimized out>) at /home/pdunaj/work/ncs/zephyr/arch/arm/core/fault.c:791
#24 0x000202b6 in __usage_fault () at /home/pdunaj/work/ncs/zephyr/arch/arm/core/fault_s.S:147
#25 <signal handler called>
#26 0x200002a0 in __kernel_ram_start ()
#27 0x2000b1a8 in sys_work_q_stack ()
```

Reply:
Tested on `b44ebc85abe48bec56e9aa0b0cf2c90b9c4bd85f`

Reply:
Darn, I think that was a type that slipped through the cracks in a rebase I did :( I will push updates asap